File: blockchain/src/main/java/pt/um/lei/masb/blockchain/TransactionOutput.java
package pt.um.lei.masb.blockchain;

import org.openjdk.jol.info.GraphLayout;
import pt.um.lei.masb.blockchain.utils.StringUtil;

import javax.persistence.*;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import java.math.BigDecimal;
import java.security.PublicKey;
import java.util.HashSet;
import java.util.Set;

@Entity
public class TransactionOutput implements Sizeable, Hashed {
    @Id
    @GeneratedValue
    private long id;

    @Basic(optional = false)
    private final PublicKey publicKey;

    @Basic(optional = false)
    private final String prevCoinbase;
    @Basic(optional = false)
    private String hashId;

    @Basic(optional = false)
    private BigDecimal payout;
    @ElementCollection
    private Set<String> tx;

    protected TransactionOutput() {
        publicKey = null;
        prevCoinbase = null;
    }

    public TransactionOutput(@NotNull PublicKey publicKey,
                             @NotNull String hashId,
                             @NotNull String prevCoinbase,
                             @NotNull BigDecimal cumUTXO,
                             @NotEmpty Set<String> tx) {
        this.publicKey = publicKey;
        this.hashId = hashId;
        this.prevCoinbase = prevCoinbase;
        payout = cumUTXO;
        this.tx = tx;
    }

    TransactionOutput(@NotNull PublicKey publicKey,
                      @NotNull String prevCoinbase,
                      @NotNull BigDecimal cumUTXO,
                      @NotNull String newT,
                      @NotNull String prev) {
        this.publicKey = publicKey;
        this.prevCoinbase = prevCoinbase;
        this.tx = new HashSet<>();
        payout = new BigDecimal("0");
        addToPayout(cumUTXO, newT, prev);
    }

    private String recalculateHash() {
        return StringUtil.getDefaultCrypter()
                         .applyHash(tx.stream()
                                      .reduce("",
                                              String::concat) + StringUtil.getStringFromKey(publicKey));
    }

    @Override
    public long getApproximateSize() {
        return GraphLayout.parseInstance(this).totalSize();
    }

    public PublicKey getPublicKey() {
        return publicKey;
    }

    public BigDecimal getPayout() {
        return payout;
    }

    public void addToPayout(@NotNull BigDecimal payout,
                            @NotNull String tx,
                            @NotNull String prev) {
        this.tx.add(prev + tx);
        this.payout = this.payout.add(payout);
        hashId = recalculateHash();
    }

    public String getPrevCoinbase() {
        return prevCoinbase;
    }

    public Set<String> getTx() {
        return tx;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String getHashId() {
        return hashId;
    }
}

File: blockchain/src/main/kotlin/pt/um/lei/masb/blockchain/TransactionOutput.kt
package pt.um.lei.masb.blockchain

import com.orientechnologies.orient.core.record.OElement
import mu.KLogging
import org.openjdk.jol.info.GraphLayout
import pt.um.lei.masb.blockchain.data.Storable
import pt.um.lei.masb.blockchain.utils.Crypter
import pt.um.lei.masb.blockchain.utils.DEFAULT_CRYPTER
import pt.um.lei.masb.blockchain.utils.Hashable
import java.math.BigDecimal
import java.security.PublicKey

data class TransactionOutput(
    val publicKey: PublicKey,
    val prevCoinbase: Hash,
    private var _hashId: Hash,
    private var payout: Payout,
    private var tx: MutableSet<Hash>
) : Sizeable, Hashed, Hashable, Storable {


    override val hashId: Hash
        get() = _hashId

    override val approximateSize: Long =
        GraphLayout.parseInstance(this)
            .totalSize()

    val payoutTX: Payout
        get() = payout

    val txSet: Set<Hash>
        get() = tx


    constructor(
        publicKey: PublicKey,
        prevCoinbase: Hash,
        cumUTXO: Payout,
        newT: Hash,
        prev: Hash
    ) : this(
        publicKey,
        prevCoinbase,
        "",
        BigDecimal("0"),
        mutableSetOf<Hash>()
    ) {
        addToPayout(cumUTXO, newT, prev)
        _hashId = digest(crypter)
    }

    override fun store(): OElement {
        TODO("not implemented")
    }

    fun addToPayout(
        payout: Payout,
        tx: Hash,
        prev: Hash
    ) {
        this.tx.add(prev + tx)
        this.payout = this.payout.add(payout)
        _hashId = digest(crypter)
    }

    /**
     * {@inheritDoc}
     */
    override fun digest(c: Crypter): Hash =
        c.applyHash(
            """
                $publicKey
                $prevCoinbase
                $hashId
                $payout
                ${tx.joinToString("") { it }}
                """.trimIndent()
        )

    override fun toString(): String =
        StringBuilder().let { sb: StringBuilder
            ->
            sb.append(
                """
                |           TransactionOutput: {
                |               PublicKey: $publicKey,
                |               PrevCoinbase: $prevCoinbase,
                |               HashId: $hashId,
                |               Payout: $payout,
                |               TXs: [
                """.trimMargin()
            )
            txSet.forEach {
                sb.append(
                    """
                    |                   $it
                    """.trimIndent()
                )
            }
            sb.append(
                """
                |               ]
                |           }
                """.trimMargin()
            )
            sb.toString()
        }

    companion object : KLogging() {
        val crypter = DEFAULT_CRYPTER
    }
}
